### PR TITLE
Implement seek 

### PR DESCRIPTION
**Pull request by cswank**
Here we go again!

This one attempts to not break any pre-existing functionality.

### CODE REVIEW

File: flac_test.go [hunk 1]
```diff
@@ -11,3 +14,108 @@ func TestSkipID3v2(t *testing.T) {
...
+		seek     int64
+		whence   int
+		expected int64
+	}{
+		{seek: 0, whence: io.SeekStart, expected: 0},
+		{seek: 9000, whence: io.SeekStart, expected: 8192},
+		{seek: 0, whence: io.SeekStart, expected: 0},
+		{seek: -6000, whence: io.SeekEnd, expected: 36864},
+
+		// expected: 40960 seems like the wrong answer, it should be
```
Comment (cswank) [inline flac_test.go:36]: What do y'all think about this?

File: flac.go [hunk 2]
```diff
@@ -64,79 +74,122 @@ func New(r io.Reader) (stream *Stream, err error) {
...
-	return block.IsLast, nil
+	return block, nil
 }
 
 // skipID3v2 skips ID3v2 data prepended to flac files.
 func (stream *Stream) skipID3v2() error {
-	r := bufio.NewReader(stream.r)
-
 	// Discard unnecessary data from the ID3v2 header.
-	if _, err := r.Discard(2); err != nil {
```
Comment (cswank) [inline flac.go:139]: These changes were the only way I could get all the tests to pass.  I don't really understand why though :-(

File: flac.go [hunk 3]
```diff
@@ -253,7 +307,139 @@ func (stream *Stream) ParseNext() (f *frame.Frame, err error) {
...
+//
+// sample is valid if:
+// whence == io.SeekEnd and sample is negative
+// whence == io.SeekStart and sample is positive
+// whence == io.SeekCurrent and sample + current sample > 0 and < stream.Info.NSamples
+//
+// If sample does not match one of the above conditions then the result will
+// probably be seeking to the beginning or very end of the data and no error
+// will be returned.
+func (stream *Stream) Seek(sample int64, whence int) (read int64, err error) {
```
[the remark this answers is not in the document]
Comment (mewmew) [reply flac.go:320]: What does the `read` return value represent? Consider documenting it. If `read == sampleNum` is always true when `err != nil`, then we can probably remove the `read` return parameter from the API without looking any power of expression.

File: flac.go [hunk 4]
```diff
@@ -253,7 +307,139 @@ func (stream *Stream) ParseNext() (f *frame.Frame, err error) {
...
+	return p, nil
+}
+
+// searchFromEnd expects sample to be negative.
+// If it is positive, it's ok, the last seek point will be returned.
+func (stream *Stream) searchFromEnd(sample int64) (p meta.SeekPoint) {
+	return stream.searchFromStart(int64(stream.Info.NSamples) + sample)
+}
+
+func (stream *Stream) searchFromStart(sample int64) (p meta.SeekPoint) {
```
Comment (mewmew) [inline flac.go:370]: Rename `sample` to `sampleNum` for consistency with variable names used in `seektable.go`.

File: flac.go [hunk 5]
```diff
@@ -253,7 +307,139 @@ func (stream *Stream) ParseNext() (f *frame.Frame, err error) {
...
+			}
+			return last
+		}
+		last = p
+	}
+	return p
+}
+
+func (stream *Stream) makeSeekTable() (err error) {
+	r, ok := stream.r.(io.ReadSeeker)
```
Comment (mewmew) [inline flac.go:386]: Rename `r` to `rs` for consistency and clarity.

**Review by mewmew — COMMENTED** [review]
A first batch of comments. Thanks a lot for making the new PR with the core functionality! It looks very clean.

File: flac.go (hunk 2 above)
Comment (cswank) [reply flac.go:139]: I reverted it back, and it works with r AND rs, so I must have had some other bug that was causing this, but seems fixed now.

**Comment by jypelle** [comment]
:+1:

**Comment by mewmew** [comment]
The PR looks good. Thanks for taking the time also to implement test cases and working through the review comments.

Really happy to finally see seek implemented in FLAC!

Cheers,
- Henry & Robin

**Comment by cswank** [comment]
Thanks, it was a lot of fun.  I'll catch you two later.

Craig

PS.  You didn't forget about [this](https://github.com/faiface/beep/pull/16#issuecomment-368826168), did you Robin?

**Comment by mewmew** [comment]
@cswank singing as we speak!

♩♪♬♩

**Comment by mewmew** [comment]
> PPS. If you wouldn't mind making a release, I'd like to make a PR for beep.

@cswank Now version [v1.0.7](https://github.com/mewkiz/flac/tree/v1.0.7) has been tagged. Would love to see seek support implemented in Beep! Looking forward to the PR :)

**Comment by cswank** [comment]
> > PPS. If you wouldn't mind making a release, I'd like to make a PR for beep.
> 
> @cswank Now version [v1.0.7](https://github.com/mewkiz/flac/tree/v1.0.7) has been tagged. Would love to see seek support implemented in Beep! Looking forward to the PR :)

Already done!  We'll see where it goes:  https://github.com/faiface/beep/pull/111

**Comment by karlek** [comment]
Haha that's awesome! Great work @cswank :grin:

**Pull request merged**